File: lib/startup-check.ts
/**
 * Production startup validation.
 *
 * Call once at app startup (e.g. from lib/scheduler.ts or a Next.js
 * instrumentation hook).  In production, missing critical env vars throw
 * immediately so the process fails loudly rather than silently misbehaving
 * at the first API call.
 *
 * In development, missing vars emit warnings but do NOT throw.
 */
import { createLogger } from './logger';

const log = createLogger('lib/startup-check');

interface CheckResult {
  passed: boolean;
  errors: string[];
  warnings: string[];
}

/**
 * REQUIRED in production — absence means a security bypass or total failure.
 */
const PRODUCTION_REQUIRED: Array<{ key: string; reason: string }> = [
  {
    key:    'ADMIN_EMAILS',
    reason: 'Without ADMIN_EMAILS every authenticated Supabase user gets full admin access.',
  },
  {
    key:    'ADMIN_SECRET',
    reason: 'Without ADMIN_SECRET the Python FastAPI backend accepts requests from anyone.',
  },
  {
    key:    'SUPABASE_SERVICE_ROLE_KEY',
    reason: 'Required for server-side database writes that bypass RLS.',
  },
  {
    key:    'NEXT_PUBLIC_SUPABASE_URL',
    reason: 'Required for all Supabase operations.',
  },
  {
    key:    'NEXT_PUBLIC_SUPABASE_ANON_KEY',
    reason: 'Required for client-side Supabase auth.',
  },
];

/**
 * RECOMMENDED — missing triggers a warning, not a fatal error.
 */
const RECOMMENDED: Array<{ key: string; reason: string }> = [
  { key: 'REDIS_URL',           reason: 'Provider metrics and caching will use in-memory fallback only.' },
  { key: 'BACKEND_URL',         reason: 'Admin panel cannot reach Python FastAPI backend.' },
  { key: 'ANTHROPIC_API_KEY',   reason: 'AI validation will use heuristic fallback for all signals.' },
  { key: 'TELEGRAM_BOT_TOKEN',  reason: 'Signal alerts and scan failure notifications are disabled.' },
  { key: 'ALLOWED_ORIGINS',     reason: 'All cross-origin API requests are denied in production.' },
];

export function runStartupCheck(): CheckResult {
  const isProduction = process.env.NODE_ENV === 'production';
  const errors: string[] = [];
  const warnings: string[] = [];

  for (const { key, reason } of PRODUCTION_REQUIRED) {
    const val = process.env[key];
    if (!val || val.trim() === '') {
      if (isProduction) {
        errors.push(`${key} is not set. ${reason}`);
      } else {
        warnings.push(`${key} is not set (dev-mode warning). ${reason}`);
      }
    }
  }

  for (const { key, reason } of RECOMMENDED) {
    const val = process.env[key];
    if (!val || val.trim() === '') {
      warnings.push(`${key} is not set. ${reason}`);
    }
  }

  const passed = errors.length === 0;

  for (const w of warnings) {
    log.warn(`startup_warning: ${w}`);
  }

  if (!passed) {
    for (const e of errors) {
      log.error(`startup_error: ${e}`);
    }
    throw new Error(
      `\n\n🚨 STARTUP FAILED — Missing required environment variables:\n` +
      errors.map(e => `  • ${e}`).join('\n') +
      `\n\nSet these in your .env.local (development) or deployment environment (production).\n`,
    );
  }

  log.info({ warnings: warnings.length }, 'startup_check_passed');
  return { passed, errors, warnings };
}

File: backend/startup_check.py
"""
Production startup validation.

Called from main.py lifespan() before the app begins accepting requests.
In production (ENVIRONMENT=production), missing critical variables raise
immediately so the process fails loudly with a clear error message rather
than silently misbehaving on the first API call.

In development, missing vars emit warnings but do NOT raise.
"""
from __future__ import annotations

import os

from backend.config import get_settings
from backend.logging.setup import get_logger

log = get_logger(__name__)

# ── Required in production ────────────────────────────────────────────────────
# (key, reason why it is critical)
_PRODUCTION_REQUIRED = [
    (
        "ADMIN_SECRET",
        "Without ADMIN_SECRET the Python FastAPI backend accepts all requests "
        "from the Next.js proxy — anyone can call admin endpoints.",
    ),
    (
        "DATABASE_URL",
        "Without DATABASE_URL asyncpg cannot connect to Postgres. "
        "All settings reads and writes will fail at runtime.",
    ),
    (
        "NEXT_PUBLIC_SUPABASE_URL",
        "Required for all Supabase SDK operations.",
    ),
    (
        "NEXT_PUBLIC_SUPABASE_ANON_KEY",
        "Required for Supabase authentication flows.",
    ),
]

# ── Recommended ───────────────────────────────────────────────────────────────
_RECOMMENDED = [
    ("REDIS_URL",           "Provider metrics and settings cache will degrade without Redis."),
    ("TELEGRAM_BOT_TOKEN",  "Scan failure alerts and signal notifications are disabled."),
    ("TELEGRAM_CHAT_ID",    "Scan failure alerts and signal notifications are disabled."),
    ("ANTHROPIC_API_KEY",   "AI validation will use heuristic fallback for all signals."),
    ("COINGECKO_API_KEY",   "CoinGecko free tier has low rate limits; provider may hit quota."),
]


def run_startup_check() -> None:
    """
    Validate the environment.  Raises RuntimeError in production if any
    required variable is missing.  In development, logs warnings only.
    """
    settings = get_settings()
    is_production = settings.is_production
    errors: list[str] = []
    warnings: list[str] = []

    for key, reason in _PRODUCTION_REQUIRED:
        value = os.environ.get(key, "").strip()
        # Also check the pydantic settings object for keys it knows about
        if not value:
            # Some keys map to pydantic field names (snake_case) — check both
            pydantic_key = key.lower()
            value = str(getattr(settings, pydantic_key, "") or "").strip()

        if not value:
            if is_production:
                errors.append(f"{key}: {reason}")
            else:
                warnings.append(f"{key} not set (dev-mode warning): {reason}")

    for key, reason in _RECOMMENDED:
        value = os.environ.get(key, "").strip()
        if not value:
            pydantic_key = key.lower()
            value = str(getattr(settings, pydantic_key, "") or "").strip()
        if not value:
            warnings.append(f"{key} not set: {reason}")

    for w in warnings:
        log.warning("startup_warning", detail=w)

    if errors:
        for e in errors:
            log.error("startup_error", detail=e)
        raise RuntimeError(
            "\n\nSTARTUP FAILED — Missing required environment variables:\n"
            + "\n".join(f"  • {e}" for e in errors)
            + "\n\nSet these in your .env.local or deployment environment.\n"
        )

    log.info("startup_check_passed", warnings=len(warnings))
